### ISSUE TITLE
Make this repo in kubernetes' image

### ISSUE THREAD

**Issue by pmorie** [kubernetes-incubator/service-catalog#79]
We have addressed this concern piecemeal in a couple different places, but I think it is advantageous to open a high-level issue for it and explain the rationale in detail.

I propose that we make the service-catalog repo, to the closest possible degree, similar to the main kubernetes repo.  This includes things like project layout, development tooling, dependency management, and other facets of developer experience.

My rationale for this is:

1.  In many, many ways, the eventual shape of the project we have discussed is very similar to Kubernetes - we are basically building a Kubernetes-like system with a smaller problem space
2.  I expect that as this project becomes known and popular, that there will be a large influx of people from the wider community; it will save all of us time and effort if someone familiar with Kubernetes at a development level can understand how to internalize and develop on service-catalog with minimal mental context-switching
3.  We are one of the first incubator repos that has the smaller-kubernetes shape, and I would like to be able to point to our repo as a how-to example for future such projects (I expect there will be very many of these once API federation is baked into how)
4.  If we want assistance from Kubernetes maintainers (as I expect that we will when we get to things like API conversion generation), it is most time efficient if they do not have to learn a new project structure
5.  Much of the tooling around code generation that we will want to pull in is very opinionated about project structure, and the path of least resistance is conforming to those opinions rather than trying to change the tooling to be more generic

With that said, I realize that this is a big ask from the group.  I have tried to make a good case for the ask on the merits, but I also want to acknowledge the fact that it is an ask.  Some reasons why it is an ask:

1.  Developers like to make projects structured the way they like and are used to, and most people in the SIG are new to the Kubernetes world to some degree
2.  Certain aspects of this ask might be alien to folks who have used go in the past in other projects / communities
3.  Certain aspects of the structure may be non-standard or have a cultural history that informs them that is not obvious, and to a great extent, the ask in this case is one of trust that it will be easiest in the future if we go this way

Let me call out special attention to (3) and emphasize this point: basically, I am asking the group to trust me when I say this way is the best one for the long-term.  Asking for trust on matters like this is something that I do not do lightly.  I understand that this is new territory for all of us -- myself included, and I want to ensure that it is known that I value and appreciate others' opinions on matters of project structure and developer experience.

Let's discuss this in a SIG meeting, no matter what discussion happens in github.  Thanks for reading.

**Comment by martinmaly** [comment]
Thanks for typing this up, Paul!

personally, I agree with much of the rationale. My primary questions are:


   - I actually am not familiar at all with how k8s does things. For that
   it would be great if someone could give a (fairly detailed) overview of k8s
   repo structure, build, tooling. Would that be possible?
   - Anticipating there are deltas between service-catalog and k8s core, we
   could then understand and enumerate them.
   - What is the level of confidence that the way k8s does things is
   because a) it is the current state of the art and they would do it the same
   way again and b) it was state of the art at some point in the past but they
   would recommend and pick different way. For b) cases - if they exist - I
   worry to pick consistency as more important than following best practices
   (while still aligned with k8s)
   - Similar to previous point - Kubernetes may be evolving their
   practices. Should we aim for k8s as it is now, or as it aims to be? One
   example of this is Bazel - I understand k8s is moving build toward Bazel.
   - Do we also adapt the k8s CI machinery, commit queues, etc.? It would
   be great to get overview of those if that's the case. (notably, I don't see
   k8s using Travis)

To summarize - in principle I agree but we don't have good understanding of
all the aspects of the alignment. I am hungry for more information and
looking forward to the meeting discussion.

Cheers, and thanks again for writing the rationale.

Martin

**Comment by pmorie** [comment]
@martinmaly 

> - I actually am not familiar at all with how k8s does things. For that it would be great if someone could give a (fairly detailed) overview of k8s repo structure, build, tooling. Would that be possible?

Repo structure is definitely doable, and I plan to document in this issue.  Build is a little more complicated.  Tooling also a bit more complicated.  There's a lot of information to convey, and some of it, we will come to when we get to it, but I will do a sketch of the level of detail we need for now.

> - Anticipating there are deltas between service-catalog and k8s core, we could then understand and enumerate them.

No doubt that there will be deltas -- but I think we can keep the same general structure in terms of repo and package layout.

>  - What is the level of confidence that the way k8s does things is because a) it is the current state of the art and they would do it the same way again and b) it was state of the art at some point in the past but they would recommend and pick different way. For b) cases - if they exist - I worry to pick consistency as more important than following best practices (while still aligned with k8s)

There is a mix of (a) and (b).  For instances of (b), there will be things we can and should do better.  There will also be things that aren't optimal but that are so strongly characteristic of k8s that we need to do them so people can easily make the jump from one to the other without a huge mental context switch.  Keeping code in `pkg/` is probably one of those -- as an example, @thockin is on record as stating that `pkg/` is not optimal, but is too costly to change now.  However, `pkg` is so characteristic of k8s that I think we should follow the same convention.

>  - Similar to previous point - Kubernetes may be evolving their practices. Should we aim for k8s as it is now, or as it aims to be? One example of this is Bazel - I understand k8s is moving build toward Bazel.

We should try to go where Kubernetes wants to, within reason.  With that said, I consider ease of context switch to be of higher importance than riding the cutting edge of desire for Kubernetes project structure.  I'm not personally familiar enough with Bazel that I currently want to adopt it.  I also don't want to get bogged down in concerns about tooling, to be blunt.  I don't want to stifle discussion, but I think we should keep in mind that the goal is to build the service-catalog, not make a project with a perfect dev experience and tooling.  None of that will matter if the thing doesn't work, and we've gotten quite far on what Kubernetes does.

>  - Do we also adapt the k8s CI machinery, commit queues, etc.? It would be great to get overview of those if that's the case. (notably, I don't see k8s using Travis) 

I would really like to, where it makes sense.  Many problems have already been solved, and at a larger level than this project, we need to be able to scale those processes as the number of repos scale.  In that dimension, part of being a good community citizen as some of the first incubator developers is to identify candidates for re-use.  With that said, we have to balance that with expediency.  I just don't want to completely ignore what has already been done and waste time re-solving problems.

**Comment by pmorie** [comment]
Main structure I would like to move to in the short term:

```
contrib/ # contributed code, not official, but things we agree are good to leave in the repo
hack/ # build, verify, munge, code generation scripts and tooling
pkg/  # source packages go here
  apis/ # api groups go here
    servicecatalog/ # service catalog API group - unversioned internal API
      v1alpha1/     # service catalog API group - versioned external API
  apiserver/   # api server code
  controller/  # controller code
  util/
test/ # test code goes here
  integration/ # integration tests
  e2e/         # e2e tests
```

**Comment by thockin** [comment]
I'll plead to avoid "hack".  Put build-related scripts in `build/` and
other tools in `tools/` or `scripts/`.

We *want* the bots to be generically useful.  Please use them and help make
that goal more real.

Re: APIs.  Our codegen tools are still k8s-centric.  I want most of them to
be more generic.  Please help.

Re Build: I posit that `make {,all,test}` is the UI to our build.  Whether
it uses Makefiles, shell scripts, gb, go install, or something else.  As
long as you stay compatible with the `make` UI it will be good.  Starting
with a hermetic dockerized build is a good thing,  See
https://github.com/thockin/go-build-template for some thoughts.  This is
one of those "skate to where the puck will be" things.

I would *not* presuppose to use ginkgo for e2e.  Consider alternatives,
please, and report back :)

**Comment by pmorie** [comment]
@thockin I would really prefer to keep this repo and kubernetes in sync from a broad structural standpoint so there is minimal context-switching required when going between one or the other.  This doesn't preclude us from making bots / other tooling more generic -- in fact, I'm sure we will have to, even if we try to closely follow kubernetes at some level.  With that said, I'm extremely wary of the pull of 'let's make this the way Kubernetes should be' and winding up in a state where we have a lot of noise / context switching overhead / mismatches because Kubernetes doesn't catch up, or goes in a different direction.

**Comment by pmorie** [comment]
So, I had a very good conversation about this subject with @martinmaly today.  One of the things that we discussed is the need to capture model of types of deltas between service catalog and Kubernetes and think through every dimension in terms of that model.  The categories of difference measurement we talked about are:

1. **Same - merits**: service-catalog and Kubernetes should be alike in a particular aspect because how Kubernetes handles that aspect has merit
2. **Same - higher purpose**: service-catalog and Kubernetes should be alike in a particular aspect which may be suboptimal or unconventional, because there is a higher purpose, such as knowledge or tooling portability which is served by the similarity
3. **Different - merit**: service-catalog should do something different than Kubernetes because the Kubernetes way is suboptimal and no higher purpose is served through similarity
4. **Doesn't matter**: service-catalog can do something different because similarity or difference is immaterial to success or failure

Let's use this model (or a better one, if anyone has suggestions!) to assess what the right thing to do in a particular area is.

**Comment by pmorie** [comment]
Let me also say explicitly that I formally renounce this comment 

> I would really prefer to keep this repo and kubernetes in sync from a broad structural standpoint so there is minimal context-switching required when going between one or the other. This doesn't preclude us from making bots / other tooling more generic -- in fact, I'm sure we will have to, even if we try to closely follow kubernetes at some level. With that said, I'm extremely wary of the pull of 'let's make this the way Kubernetes should be' and winding up in a state where we have a lot of noise / context switching overhead / mismatches because Kubernetes doesn't catch up, or goes in a different direction.

...in favor of reassessing any area where there is a desire for it in terms of the model above:

cc @thockin @kubernetes-incubator/maintainers-service-catalog

**Comment by duglin** [comment]
Please don't interpret this as pushing back on the desire of this issue, but I do feel like its worth mentioning that trying to force one way of doing things on projects is often a religious battle and not worth the effort. Again, not suggesting we shouldn't align with Kube - we are after all a Kube project.  However, any good project's structure/process/tooling/etc.. (no matter what they choose) should be easy enough that a newbie does not get lost, overwhelmed and leaves because of the pain of it.

So, with that, let me rephrase it in even more blunt terms :-)  If a Kube developer comes to our repo and can't figure out our modified Kube layout/process/tooling/.... then we messed up.  Likewise, if a Kube newbie comes to our repo and can't figure it out.... then we messed up.  So, this doesn't mean blindly following or ignoring Kube's chosen path - its about making on-boarding as easy as possible for both.

I'm ok with diverging from Kube if it improves things in a way that's logical to both newbies and existing Kube developers.  I have no idea where that line is but I'm willing to look at things on a case-by-case basis. And I think its very fair for us to ask the Kube experts to explain why things are the way they are since sometimes things slowly morph into a less than ideal state w/o people realizing it. And its good to step back sometimes and look at things with a fresh pair of eyes.

**Comment by pmorie** [comment]
I added the babynetes label because I want to capture that this is part of the challenge of babynetes.  Devs gonna dev in the absense of a template or coherent explanatory doc.  If we had such a thing to refer to, this entire experience would have been moot, or at least far more focused and less exhausting.

**Comment by pmorie** [comment]
We've made a lot of progress here; there's more to do, but I don't think we need to keep this issue open any longer.  We can reopen if necessary.
